pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
         )
         results = self._inner_execute(jac_tapes)
         return tuple(batch_post_processing(results))
+
+
+class DeviceJacobians(JacobianProductCalculator):
+    """Calculate jacobian products via a device.
+
+    Args:
+        device (Union[pennylane.Device, pennylane.devicesDevice]): the device for execution and derivatives.
+            Must support supports first order gradients with the requested configuration.
+        gradient_kwargs (dict): a dictionary of keyword options for the gradients. Only used with a :class:`~.pennylane.Device`
+            old device interface.
+        execution_config (pennylane.devices.ExecutionConfig): a datastructure containing the parameters needed to fully
+           describe the execution. Only used with :class:`pennylane.devices.Device` new device interface.
+
+    **Examples:**
+
+    >>> device = qml.device('default.qubit')
+    >>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
+    >>> jpc = DeviceJacobians(device, {}, config)
+
+    This same class can also be used with the old device interface.
+
+    >>> device = qml.device('lightning.qubit', wires=5)
+    >>> gradient_kwargs = {"method": "adjoint_jacobian"}
+    >>> jpc_lightning = DeviceJacobians(device, gradient_kwargs)
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and jacobian for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents, the
+    caching and retrieval should be more performant than a lookup based on the potentially expensive ``QuantumScript.hash``.
```

Comment:
I'm not sure I follow here. The loookup _is_ based on `QuantumScript.hash`, which is based on the object location (which is not expensive, is it?)? If not, what is it based on?

Reply:
I'm trying to specify that we are *not* using `QuantumScript.hash`, as that is expensive and not worth it. So you could have two tapes that are "identical", but have different locations in memory and will be separate caches.